server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -431,26 +439,52 @@ public void onResponse(RolloverResponse result) {
 
                 @Override
                 public void onFailure(Exception e) {
-                    failRequestsWhenPrerequisiteActionFailed(dataStream, bulkRequest, responses, e);
+                    dataStreamExceptions.put(dataStream, e);
                 }
             }, refs.acquire()));
         }
     }
 
+    // Separate method to allow for overriding in tests.
+    void rollOver(RolloverRequest rolloverRequest, ActionListener<RolloverResponse> listener) {
+        rolloverClient.execute(LazyRolloverAction.INSTANCE, rolloverRequest, listener);
+    }
+
     /**
-     * Fails all requests involving this index or data stream because the prerequisite action failed too.
+     * Mark all the requests for which the prerequisite action failed (i.e. index creation or data stream/failure store rollover) as failed.
      */
-    private static void failRequestsWhenPrerequisiteActionFailed(
-        String target,
+    private void failRequestsWhenPrerequisiteActionFailed(
+        Map<String, Exception> indicesExceptions,
+        Map<String, Exception> dataStreamExceptions,
+        Map<String, Exception> failureStoreExceptions,
         BulkRequest bulkRequest,
-        AtomicArray<BulkItemResponse> responses,
-        Exception error
+        AtomicArray<BulkItemResponse> responses
     ) {
+        if (indicesExceptions.isEmpty() && dataStreamExceptions.isEmpty() && failureStoreExceptions.isEmpty()) {
+            return;
+        }
         for (int i = 0; i < bulkRequest.requests.size(); i++) {
             DocWriteRequest<?> request = bulkRequest.requests.get(i);
-            if (request != null && setResponseFailureIfIndexMatches(responses, i, request, target, error)) {
-                bulkRequest.requests.set(i, null);
+            if (request == null) {
+                continue;
             }
+            var exception = indicesExceptions.get(request.index());
+            if (exception == null) {
+                if (request instanceof IndexRequest indexRequest && indexRequest.isWriteToFailureStore()) {
+                    exception = failureStoreExceptions.get(request.index());
+                } else {
+                    exception = dataStreamExceptions.get(request.index());
+                }
+            }
+            if (exception == null) {
+                continue;
+            }
+            var failureStoreStatus = request instanceof IndexRequest ir && ir.isWriteToFailureStore()
+                ? IndexDocFailureStoreStatus.FAILED
+                : IndexDocFailureStoreStatus.NOT_APPLICABLE_OR_UNKNOWN;
+            var failure = new BulkItemResponse.Failure(request.index(), request.id(), exception, failureStoreStatus);
```

Comment:
@jbaiera I copied this from
https://github.com/elastic/elasticsearch/blob/f7ba89bcb1f7f059542dcccd83c01d06ed613dfa/server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java#L765-L768
(as per https://github.com/elastic/elasticsearch/pull/112816)
Can you quickly re-check these changes with that merged PR in mind? I think everything is ok, but would like you to have a quick check to be sure.